trunk/source/Library-Proofs/src/de/uni_freiburg/informatik/ultimate/lib/proofs/floydhoare/FloydHoareMapping.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2023 Dominik Klumpp (klumpp@informatik.uni-freiburg.de)
+ * Copyright (C) 2023 University of Freiburg
+ *
+ * This file is part of the ULTIMATE Proofs Library.
+ *
+ * The ULTIMATE Proofs Library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ULTIMATE Proofs Library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with the ULTIMATE Proofs Library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Additional permission under GNU GPL version 3 section 7:
+ * If you modify the ULTIMATE Proofs Library, or any covered work, by linking
+ * or combining it with Eclipse RCP (or a modified version of Eclipse RCP),
+ * containing parts covered by the terms of the Eclipse Public License, the
+ * licensors of the ULTIMATE Proofs Library grant you additional permission
+ * to convey the resulting work.
+ */
+package de.uni_freiburg.informatik.ultimate.lib.proofs.floydhoare;
+
+import java.util.Map;
+
+import de.uni_freiburg.informatik.ultimate.lib.modelcheckerutils.smt.predicates.IPredicate;
+import de.uni_freiburg.informatik.ultimate.lib.proofs.PrePostConditionSpecification;
+
+/**
+ * A simple implementation of {@link IFloydHoareAnnotation} backed by a map.
+ *
+ * @author Dominik Klumpp (klumpp@informatik.uni-freiburg.de)
+ *
+ * @param <S>
+ *            the type of states which are annotated
+ */
+public class FloydHoareMapping<S> implements IFloydHoareAnnotation<S> {
+	private final PrePostConditionSpecification<S> mSpecification;
+	private final Map<S, IPredicate> mAnnotation;
+	private final IPredicate mDefaultPredicate;
+
+	public FloydHoareMapping(final PrePostConditionSpecification<S> specification,
```

Comment:
We use `null` to indicate that a location does not have any annotation. Since we have settings to only compute the annotation at certain locations (e.g. loop heads), this can be the case for ICFGs; and there is no non-null default predicate we could use in this case.

Reply:
Its just that having no annotation would mean I could always use true in that case, right?

Reply:
No, that would be different. Imagine an edge `src --(x:=0)--> tgt`, where `tgt` has an annotation like `x==y`. If we check the validity of this annotation (as in `FloydHoareValidityChecker`), we have to differentiate:
- If `src` has the annotation `true` (perhaps as default predicate), the Hoare triple `{true} x:=0 {x==y}` is not valid, and so the entire annotation is invalid.
- If `src` has the annotation `null`, it is clear that we do not have the full Hoare triple to check. But we don't consider the annotation invalid in this case.

Reply:
Ah, perhaps this should be a comment for this class :)

Reply:
:+1: I've added this explanation to `IFloydHoareAnnotation` and linked to it in the documentation for the two different constructors here.